sycl/plugins/unified_runtime/ur/adapters/cuda/adapter.cpp

```diff
@@ -0,0 +1,82 @@
+//===--------- adapter.cpp - CUDA Adapter ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===-----------------------------------------------------------------===//
+
+#include <ur_api.h>
+
+#include "common.hpp"
+
+void enableCUDATracing();
+void disableCUDATracing();
+
+struct ur_adapter_handle_t_ {
+  std::atomic<uint32_t> RefCount = 0;
+};
+
+ur_adapter_handle_t_ adapter{};
+
+UR_APIEXPORT ur_result_t UR_APICALL urInit(ur_device_init_flags_t,
+                                           ur_loader_config_handle_t) {
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL urTearDown(void *) {
+  return UR_RESULT_SUCCESS;
+}
+
+UR_APIEXPORT ur_result_t UR_APICALL
+urAdapterGet(uint32_t NumEntries, ur_adapter_handle_t *phAdapters,
+             uint32_t *pNumAdapters) {
+  if (phAdapters) {
+    if (adapter.RefCount++ == 0) {
+      enableCUDATracing();
```

Comment:
This might be a race where threads that loaded `RefCount >= 1` could see an uninitialized state if the first thread to enter this condition hasn't finished.

Reply:
Good point, I've added a lock to this and the L0 setup/teardown to prevent this